### PR TITLE
 Make dtype consistent in filters.rank functions

### PR DESCRIPTION
**Pull request by hmaarrfk**
## Description

    Added default arg to selem to make consistent with docs
    Removed default_selem decorator and bool image dtype support
    Made selem=None default argument for all rank functions

Closes #3981 @clementkng @sciunto @lagru @soupault

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by hmaarrfk** [comment]
As a reference this is what I did to rebase:

1. Create a new branch
1. Use `git log` to find 1 commit before the one you authored
2. Rebase and squash onto that one commit.

```
git rebase -i 3a139607d642211bf681f4266b426b355becd099
```
removing the merge commits, and squashing as many commits as possible. This avoids repeated failed 3 way merges in the case of conflicts.
3. Update master
4. `git rebase -i master`

tada

**Comment by jni** [comment]
@hmaarrfk that's a *fantastic* user guide for hairy rebases.

**Comment by jni** [comment]
@hmaarrfk would you mind updating the PR description with a bit more motivation for each change? It would also be good to specify that `default_selem` is removed only for the median filter.

**Comment by hmaarrfk** [comment]
Oh man, i don't even remember what this PR is about. I just went on a git binge....

File: skimage/filters/rank/generic.py [hunk 1]
```diff
@@ -23,13 +23,12 @@
 image) where non zero values are the part of the image participating in the
 histogram computation. By default the entire image is filtered.
 
-This implementation outperforms grey.dilation for large structuring elements.
+This implementation outperforms gray.dilation for large structuring elements.
```
Comment (stefanv) [inline skimage/filters/rank/generic.py]: What does `gray.dilation` refer to here?

File: skimage/filters/rank/generic.py [hunk 2]
```diff
@@ -23,13 +23,12 @@
...
 
-Input image can be 8-bit or 16-bit, for 16-bit input images, the number of
-histogram bins is determined from the maximum value present in the image.
-
-Result image is 8-/16-bit or double with respect to the input image and the
-rank filter operation.
+Input images will be cast in unsigned 8-bit integer or unsigned 16-bit integer
+if necessary. The number of histogram bins is then determined from the maximum
+value present in the image. Eventually, the output image is cast in the desired
```
Comment (stefanv) [inline skimage/filters/rank/generic.py]: Unclear what the "desired dtype" would be here.

File: skimage/filters/rank/generic.py [hunk 3]
```diff
@@ -56,22 +55,59 @@
...
 
-
 __all__ = ['autolevel', 'bottomhat', 'equalize', 'gradient', 'maximum', 'mean',
            'geometric_mean', 'subtract_mean', 'median', 'minimum', 'modal',
            'enhance_contrast', 'pop', 'threshold', 'tophat', 'noise_filter',
            'entropy', 'otsu']
 
 
-def _handle_input(image, selem, out, mask, out_dtype=None, pixel_size=1):
+def _handle_input(image, selem=None, out=None, mask=None, out_dtype=None,
```
Comment (stefanv) [inline skimage/filters/rank/generic.py]: Not sure how we originally came up with `_handle_input`; seems like a pretty terrible name! Can we be more descriptive?

File: skimage/filters/rank/generic.py (hunk 1 above)
Comment (emmanuelle) [reply skimage/filters/rank/generic.py]: Maybe `gray_dilation`?

Comment (jni) [reply skimage/filters/rank/generic.py]: Actually this refers to the `dilation` function in `grey.py`, so the new wording is incorrect, unfortunately.

Comment (emmanuelle) [reply skimage/filters/rank/generic.py]: Probably an outdated comment, meaning `morphology.dilation`.

**Comment by emmanuelle** [comment]
@hmaarrfk if you're ok with that, I'll push a commit with the minor changes requested by @stefanv and merge so that we can benefit from the improvements of this PR.

**Pull request merged**